backends/bmv2/common/parser.cpp

```diff
@@ -106,11 +143,17 @@ Util::IJson *ParserConverter::convertParserStatement(const IR::StatOrDecl *stat)
     }
     if (stat->is<IR::AssignmentStatement>()) {
         auto assign = stat->to<IR::AssignmentStatement>();
-        auto type = ctxt->typeMap->getType(assign->left, true);
-        cstring operation = jsonAssignment(type, true);
+        auto expr = assign->left;
+        auto type = ctxt->typeMap->getType(expr, true);
+        cstring operation = jsonAssignment(type, expr, true);
         result->emplace("op", operation);
-        auto l = ctxt->conv->convertLeftValue(assign->left);
         bool convertBool = type->is<IR::Type_Boolean>();
+        Util::IJson *l;
+        if (ctxt->conv->isArrayIndexRuntime(assign->left)) {
```

Comment:
I am not very familiar with the bmv2 backend code, maybe it would be nice to have a comment to explain why this is handled as a special case?

Shouldn't `convertLeftValue` also be able to handle this case gracefully?

Reply:
Sure, will do. Just want to give a short explanation here before I have time to address other comments. This part was added to address the issue described in https://github.com/p4lang/p4c/issues/4194#issuecomment-1763706371. One observation I had was that the issue only occurs in parser block, but not control block. So I looked at the corresponding code for control block (see below), and basically copied the diff here without fully understanding why it works.

https://github.com/p4lang/p4c/blob/ac40e1f0812fac8e5d0744bc7febb4c1fb0e6a1e/backends/bmv2/common/action.cpp#L122-L126

Reply:
I've dug up where this logic came from: https://github.com/p4lang/p4c/pull/2139/files#r369298969.

To quote @hesingh:
> It is needed to deal with the case when a runtime index is a left value and is the only expression in JSON (without any other arithmetic operations). The JSON format needs an extra wrapping `"type": "expression"`.
> 
> Regarding convertLeftValue, note, the right value is also using `convert` because simple_switch supports runtime index in right value with simple_switch. See this code few lines below the left code.
> 
> `auto right = ctxt->conv->convert(r, true, true, convertBool);`

Reply:
I've added this explanation as comments in the code.